ostd/src/logger.rs

```diff
@@ -10,96 +10,74 @@
 //!
 //! IRQs are disabled while printing. So do not print long log messages.
 
+use core::str::FromStr;
+
 use log::{LevelFilter, Metadata, Record};
+use spin::Once;
+
+static LOGGER: Logger = Logger::new();
 
-use crate::{
-    boot::{kcmdline::ModuleArg, kernel_cmdline},
-    timer::Jiffies,
-};
+struct Logger {
+    backend: Once<&'static dyn log::Log>,
```

Comment:
I'm a bit confused about this implementation. Why not have the OSTD user call `log::set_logger`directly?

This `logger` is already a backend for the `log` crate, and the current changes that add the backend's backend are kind of weird and redundant (as you can see [here](https://docs.rs/log/0.4.22/src/log/lib.rs.html#1498-1513), the `log` crate already writes the same logic, although it doesn't rely on `spin::Once`).

I think we can just move the entire file from OSTD to the Asterinas kernel. This is because:
 - The whole file can be in Asterinas kernel as it doesn't contain any unsafe code.
 - Typically, `log::set_logger` should be called by an executable target instead of a library target, as [documented](https://docs.rs/log/latest/log/index.html#usage) by the `log` crate (quoted below).
   - In our scenarios, OSTD is just a library so it shouldn't call `log::set_logger`, instead, the Asterinas kernel is an executable which should be responsible to choose the logger implementation.
> Libraries should link only to the log crate, and use the provided macros to log whatever information will be useful to downstream consumers.
> [ .. ]
> Executables should choose a logging implementation and initialize it early in the runtime of the program. Logging implementations will typically include a function to do this. Any log messages generated before the implementation is initialized will be ignored.

The only downside of the change suggested above is that OSTD will not have a default logger, but I think this is reasonable since all libraries should _not_ have a logger implementation (as justified by the comments above).

Even if for some reason we need to keep a default logger in OSTD, we can still have the OSTD user call `log::set_logger` first, then the `log::set_logger` with the default logger in OSTD will fail with some [`SetLoggerError`](https://docs.rs/log/latest/log/struct.SetLoggerError.html). This is still fine and would have exactly the same effect as the current implementation in this PR.

Reply:
> Why not have the OSTD user call log::set_loggerdirectly?

This is a good question that desires a good answer.

Reply:
> Why not have the OSTD user call log::set_loggerdirectly?

This is because there are log output during the initialization of OSTD (e.g. in allocator::init()). such log operations will be executed before any initialization of components or modules in kernel services, as a result, if not setting a logger in OSTD first, we will lost these logs.

> but I think this is reasonable since all libraries should not have a logger implementation

I think this makes sense, but currently, as mentioned above, if we do this we might need to remove the log operations during the OSTD init process and directly use `early_print` instead?

Reply:
When implementing the IoPortDispatcher (#1518), I found that the files `ostd::arch::x86::device::serial.rs` and `ostd::arch::x86::serial.rs` can be removed if we disable logging in OSTD. Similarly, equivalent code in RISC-V can also be removed. 

This change could help cut down around 250 lines of code in OSTD.

Reply:
> This is because there are log output during the initialization of OSTD (e.g. in allocator::init()). such log operations will be executed before any initialization of components or modules in kernel services, as a result, if not setting a logger in OSTD first, we will lost these logs.

Well, the explanation sounds reasonable to me. I think the current implementation is at least an acceptable workaround.

We may consider disallowing the use of `log` during OSTD initialization as future work. That also sounds right to me, but I don't really know if it's necessary at the moment.